Add Check Support for Java 21 Pattern Matching for Switch Syntax: FallThrough

child of https://github.com/checkstyle/checkstyle/issues/14961

I have read check documentation:https://checkstyle.org/checks/coding/fallthrough.html#FallThrough
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="FallThrough">

        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test {

    void test(Object obj) {
        switch (obj) {
            case ColoredPoint(int x,_,_) when (x >= 2):
                System.out.println("coloredPoint");
            case Rectangle(_,_) : // violation
                System.out.println("Rectangle");
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + obj);
        }
         switch (obj) {
            case ColoredPoint(int x,_,_) when (x >= 2):
                System.out.println("coloredPoint");
                break;
            case Rectangle(var y,_) :
                System.out.println("Rectangle");
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + obj);
        }
        switch (obj) {
            case ColoredPoint(int x,_,_) when (x >= 2) ->
                System.out.println("coloredPoint");
            case Rectangle(var y,_) ->
                System.out.println("Rectangle");
            default->
                throw new IllegalStateException("Unexpected value: " + obj);
        }
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:13: Fall through from previous branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 1 errors.
```

Describe what you want in detail

we are ok here. we can close or approve for input file. 
note that:
- the switch rule (arrow syntax) isn't subject to fall through by design
- fall through to a case label declaring a pattern will not compile 
     Example:
 ```
       switch (obj) {
            case ColoredPoint(int x,_,_) when (x >= 2):
                System.out.println("coloredPoint");
            case Rectangle(var y,_) :       // compile-time error Illegal fall-through to a pattern
                System.out.println("Rectangle");
                break;
            default:
                throw new IllegalStateException("Unexpected value: " + obj);
        }
```

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheckTest.java
@@ -391,4 +391,37 @@ public class FallThroughCheckTest extends AbstractModuleTestSupport {
                 getNonCompilablePath("InputFallThroughSwitchRules.java"),
                 expected);
     }
+
+    @Test
+    public void testFallThroughWithPatternMatching() throws Exception {
+        final String[] expected = {
+            "19:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "22:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "31:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "35:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "44:13: " + getCheckMessage(MSG_FALL_THROUGH),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFallThroughWithPatternMatching.java"),
+                expected);
+    }
+
+    @Test
+    public void testFallThroughWithPatternMatchingCheckLastCase() throws Exception {
+        final String[] expected = {
+            "19:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "21:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "21:13: " + getCheckMessage(MSG_FALL_THROUGH_LAST),
+            "33:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "36:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "36:13: " + getCheckMessage(MSG_FALL_THROUGH_LAST),
+            "48:13: " + getCheckMessage(MSG_FALL_THROUGH),
+            "61:13: " + getCheckMessage(MSG_FALL_THROUGH_LAST),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputFallThroughWithPatternMatchingCheckLastCase.java"),
+                expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithPatternMatching.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithPatternMatching.java
@@ -1,0 +1,95 @@
+/*
+FallThrough
+checkLastCaseGroup = (default)false
+reliefPattern = (default)falls?[ -]?thr(u|ough)
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.fallthrough;
+
+public class InputFallThroughWithPatternMatching {
+
+    void testSwitchStatements(Object o) {
+        switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty():
+                System.out.println("empty string");
+            // violation below, 'Fall\ through from previous branch of the switch'
+            default:
+                System.out.println("default");
+        }
+
+        switch (o) {
+            case Integer i when i > 0: {
+                System.out.println("Positive integer");
+            }
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty(): {
+                System.out.println("empty string");
+            }
+            // violation below, 'Fall\ through from previous branch of the switch'
+            default: {
+                System.out.println("default");
+            }
+        }
+
+        int x = switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty():
+                yield 2;
+            default:
+                yield 3;
+        };
+
+        switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+                break;
+            case String s when o.toString().isEmpty():
+                System.out.println("empty string");
+                break;
+            default:
+                System.out.println("default");
+        }
+    }
+
+    void testSwitchRule(Object o) {
+        switch (o) {
+            case Integer i when i > 0 -> System.out.println("Positive integer");
+            case String s when s.isEmpty() -> System.out.println("empty string");
+            default -> System.out.println("default");
+        }
+
+        switch (o) {
+            case Integer i when i > 0 -> {
+                System.out.println("Positive integer");
+            }
+            case String s when s.isEmpty() -> {
+                System.out.println("empty string");
+            }
+            default -> System.out.println("default");
+        }
+
+        int x = switch (o) {
+            case Integer i when i > 0 -> {
+                System.out.println("Positive integer");
+                yield 2;
+            }
+            case String s when s.isEmpty() -> {
+                System.out.println("empty string");
+                yield 3;
+            }
+            default -> {
+                System.out.println("default");
+                yield 4;
+            }
+        };
+
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithPatternMatchingCheckLastCase.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThroughWithPatternMatchingCheckLastCase.java
@@ -1,0 +1,99 @@
+/*
+FallThrough
+checkLastCaseGroup = true
+reliefPattern = (default)falls?[ -]?thr(u|ough)
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.fallthrough;
+
+public class InputFallThroughWithPatternMatchingCheckLastCase {
+
+    void testSwitchStatements(Object o) {
+        switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty():
+                System.out.println("empty string");
+            default:
+            // 2 violations above:
+            //                     'Fall\ through from previous branch of the switch'
+            //                     'Fall\ through from the last branch of the switch'
+                System.out.println("default");
+        }
+
+        switch (o) {
+            case Integer i when i > 0: {
+                System.out.println("Positive integer");
+            }
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty(): {
+                System.out.println("empty string");
+            }
+            default: {
+            // 2 violations above:
+            //                     'Fall\ through from previous branch of the switch'
+            //                     'Fall\ through from the last branch of the switch'
+                System.out.println("default");
+            }
+        }
+
+        int x = switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+            // violation below, 'Fall\ through from previous branch of the switch'
+            case String _ when o.toString().isEmpty():
+                yield 2;
+            default:
+                yield 3;
+        };
+
+        switch (o) {
+            case Integer i when i > 0:
+                System.out.println("Positive integer");
+                break;
+            case String s when o.toString().isEmpty():
+                System.out.println("empty string");
+                break;
+            default:   // violation 'Fall\ through from the last branch of the switch'
+                System.out.println("default");
+        }
+    }
+
+    void testSwitchRule(Object o) {
+        switch (o) {
+            case Integer i when i > 0 -> System.out.println("Positive integer");
+            case String s when s.isEmpty() -> System.out.println("empty string");
+            default -> System.out.println("default");
+        }
+
+        switch (o) {
+            case Integer i when i > 0 -> {
+                System.out.println("Positive integer");
+            }
+            case String s when s.isEmpty() -> {
+                System.out.println("empty string");
+            }
+            default -> System.out.println("default");
+        }
+
+        int x = switch (o) {
+            case Integer i when i > 0 -> {
+                System.out.println("Positive integer");
+                yield 2;
+            }
+            case String s when s.isEmpty() -> {
+                System.out.println("empty string");
+                yield 3;
+            }
+            default -> {
+                System.out.println("default");
+                yield 4;
+            }
+        };
+
+    }
+}

